### PR TITLE
Integrate Soroban Domain

### PR DESCRIPTION
**Pull request by tupui**
Trying to get Soroban Domain registration in the registration itself.

https://sorobandomains.org

An alternative would be to only check that the project is register. We would call from the frontend the Soroban Domain contract or redirect to the official dApp.

### CODE REVIEW

File: contracts/versioning/src/lib.rs [hunk 1]
```diff
@@ -97,6 +99,22 @@ impl Versioning {
             panic_with_error!(&env, &ContractErrors::ProjectAlreadyExist);
         } else {
             auth_maintainers(&env, &maintainer, &project.maintainers);
+
+            let contract_domain_id_str = String::from_str(&env, CONTRACT_DOMAIN_ID);
+            let contract_domain_id = Address::from_string(&contract_domain_id_str);
+            let tld = Bytes::from_slice(&env, &[120, 108, 109]); // xlm
+            env.invoke_contract::<()>(
```
Comment (tupui) [inline contracts/versioning/src/lib.rs:106]: @earrietadev hi 👋 if you have time, could you help us out here? I am getting a failure calling Soroban Domain's contract here and I don't understand the backtrace. Thanks in advance!

```
HostError: Error(WasmVm, InvalidAction)

Event log (newest first):
   0: [Diagnostic Event] topics:[error, Error(WasmVm, InvalidAction)], data:"escalating error to panic"
   1: [Diagnostic Event] topics:[error, Error(WasmVm, InvalidAction)], data:["contract call failed", register, [CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, "soroban-versioning", [CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOLZM], "github.com/file.toml", "2ef4f49fdd8fa9dc463f1f06a094c26b88710990"]]
   2: [Failed Diagnostic Event (not emitted)] contract:CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAK3IM, topics:[error, Error(WasmVm, InvalidAction)], data:"caught error from function"
   3: [Failed Diagnostic Event (not emitted)] contract:CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAK3IM, topics:[error, Error(WasmVm, InvalidAction)], data:"escalating error to panic"
   4: [Failed Diagnostic Event (not emitted)] contract:CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAK3IM, topics:[error, Error(WasmVm, InvalidAction)], data:["contract call failed", set_record, [Bytes(736f726f62616e2d76657273696f6e696e67), Bytes(786c6d), CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, 31536000]]
   5: [Failed Diagnostic Event (not emitted)] contract:CATRNPHYKNXAPNLHEYH55REB6YSAJLGCPA4YM6L3WUKSZOPI77M2UMKI, topics:[error, Error(WasmVm, InvalidAction)], data:["VM call trapped: UnreachableCodeReached", set_record]
   6: [Failed Diagnostic Event (not emitted)] contract:CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAK3IM, topics:[fn_call, Bytes(2716bcf8536e07b567260fdec481f62404acc2783986797bb5152cb9e8ffd9aa), set_record], data:[Bytes(736f726f62616e2d76657273696f6e696e67), Bytes(786c6d), CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, 31536000]
   7: [Diagnostic Event] topics:[fn_call, Bytes(0000000000000000000000000000000000000000000000000000000000000005), register], data:[CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, "soroban-versioning", [CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOLZM], "github.com/file.toml", "2ef4f49fdd8fa9dc463f1f06a094c26b88710990"]
   8: [Diagnostic Event] contract:CBUSYNQKASUYFWYC3M2GUEDMX4AIVWPALDBYJPNK6554BREHTGZ2IUNF, topics:[fn_return, mint], data:Void
   9: [Contract Event] contract:CBUSYNQKASUYFWYC3M2GUEDMX4AIVWPALDBYJPNK6554BREHTGZ2IUNF, topics:[mint, CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAFCT4, CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, "aaa:GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGO6V"], data:10000000000000000
   10: [Diagnostic Event] topics:[fn_call, Bytes(692c360a04a982db02db346a106cbf008ad9e058c384bdaaf77bc0c48799b3a4), mint], data:[CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMDR4, 10000000000000000]
   11: [Diagnostic Event] contract:CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAK3IM, topics:[fn_return, init], data:Void
   12: [Diagnostic Event] topics:[fn_call, Bytes(0000000000000000000000000000000000000000000000000000000000000005), init], data:CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAITA4
   13: [Diagnostic Event] contract:CATRNPHYKNXAPNLHEYH55REB6YSAJLGCPA4YM6L3WUKSZOPI77M2UMKI, topics:[fn_return, init], data:Void
   14: [Diagnostic Event] topics:[fn_call, Bytes(2716bcf8536e07b567260fdec481f62404acc2783986797bb5152cb9e8ffd9aa), init], data:[CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD2KM, 100, CBUSYNQKASUYFWYC3M2GUEDMX4AIVWPALDBYJPNK6554BREHTGZ2IUNF, 31536000, [Bytes(786c6d), Bytes(7374656c6c6172), Bytes(77616c6c6574), Bytes(64616f)]]
   15: [Diagnostic Event] contract:CBUSYNQKASUYFWYC3M2GUEDMX4AIVWPALDBYJPNK6554BREHTGZ2IUNF, topics:[fn_return, set_admin], data:Void
   16: [Contract Event] contract:CBUSYNQKASUYFWYC3M2GUEDMX4AIVWPALDBYJPNK6554BREHTGZ2IUNF, topics:[set_admin, GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGO6V, "aaa:GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGO6V"], data:CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAFCT4
   17: [Diagnostic Event] topics:[fn_call, Bytes(692c360a04a982db02db346a106cbf008ad9e058c384bdaaf77bc0c48799b3a4), set_admin], data:CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAFCT4
   18: [Diagnostic Event] contract:CBUSYNQKASUYFWYC3M2GUEDMX4AIVWPALDBYJPNK6554BREHTGZ2IUNF, topics:[fn_return, init_asset], data:Void
   19: [Diagnostic Event] topics:[fn_call, Bytes(692c360a04a982db02db346a106cbf008ad9e058c384bdaaf77bc0c48799b3a4), init_asset], data:Bytes(0000000161616100000000000000000000000000000000000000000000000000000000000000000000000003)
```

File: contracts/versioning/src/test.rs [hunk 2]
```diff
@@ -1,17 +1,56 @@
...
 #[test]
 fn test() {
     let env = Env::default();
     env.mock_all_auths();
 
-    let contract_admin = Address::generate(&env);
+    // setup for Soroban Domain
+    let contract_domain_id_str = String::from_str(&env, CONTRACT_DOMAIN_ID);
+    let contract_domain_id = Address::from_string(&contract_domain_id_str);
+    env.register_contract_wasm(&contract_domain_id, contract_domain::WASM);
```
Comment (tupui) [inline contracts/versioning/src/test.rs:24]: Registering here with the same address as mainnet

File: contracts/versioning/src/lib.rs (hunk 1 above)
Comment (earrietadev) [reply contracts/versioning/src/lib.rs:106]: @tupui doing a quick review, one thing I noticed is that you're hashing the domain before passing it to the contract so you need to make sure the value is not higher than the limit which is 15 digits. Also the last value you're passing is a i32 but the contract is expecting a u64

Comment (tupui) [reply contracts/versioning/src/lib.rs:106]: Thanks I really appreciate 🙏 I will try this out!

File: contracts/versioning/src/lib.rs [hunk 3]
```diff
@@ -97,6 +99,22 @@ impl Versioning {
...
+                &contract_domain_id,
+                &Symbol::new(&env, "set_record"),
+                vec![
+                    &env,
+                    name_b.to_val(),
+                    tld.to_val(),
+                    maintainer.to_val(),
+                    maintainer.to_val(),
+                    Val::from_i32(31536000).into(),
```
Comment (tupui) [inline contracts/versioning/src/lib.rs:115]:
```suggestion
                    Val::from_u64(31536000).into(),
```

File: contracts/versioning/src/lib.rs [hunk 4]
```diff
@@ -97,6 +99,22 @@ impl Versioning {
...
+            env.invoke_contract::<()>(
+                &contract_domain_id,
+                &Symbol::new(&env, "set_record"),
+                vec![
+                    &env,
+                    name_b.to_val(),
+                    tld.to_val(),
+                    maintainer.to_val(),
+                    maintainer.to_val(),
+                    Val::from_u64(31536000).into(),
```
Comment (tupui) [inline contracts/versioning/src/lib.rs:115]:
```suggestion
                    Val::from(31536000).into(),
```

**Comment by tupui** [comment]
Ok, seems like I got it to work now, thanks again @earrietadev  😃 ! It was hard to debug and in the end was just a matter of formatting. The tooling needs to improve a bit to properly propagate errors coming from other contract calls 😅 

No I can play and see if we actually want to do this from the contract itself or just check that a domain was registered before and belongs to the person calling our register function.

**Pull request merged**